python/ray/util/spark/cluster_init.py

```diff
@@ -0,0 +1,684 @@
+                    "call `ray_cluster_on_spark.shutdown()` before detaching your databricks "
+                    "python REPL."
+                )
+        return ray_cluster_handler
+    except Exception:
+        # If driver side setup ray-cluster routine raises exception, it might result in part of ray
+        # processes has been launched (e.g. ray head or some ray workers have been launched),
+        # calling `ray_cluster_handler.shutdown()` to kill them and clean status.
+        ray_cluster_handler.shutdown()
+        raise
+
+
+_active_ray_cluster = None
+
+
+@PublicAPI(stability="alpha")
+def init_ray_cluster(
+    num_worker_nodes=None,
+    total_cpus=None,
+    total_gpus=None,
+    total_heap_memory_bytes=None,
+    total_object_store_memory_bytes=None,
+    heap_to_object_store_memory_ratio=None,
+    head_options=None,
+    worker_options=None,
+    ray_temp_root_dir="/tmp",
+    ray_log_root_dir="/tmp",
+    safe_mode=True,
+):
+    """
+    Initialize a ray cluster on the spark cluster by starting a ray head node in the spark
+    application's driver side node.
+    After creating the head node, a background spark job is created that
+    generates an instance of `RayClusterOnSpark` that contains configuration for the ray cluster
+    that will run on the Spark cluster's worker nodes.
+    After a ray cluster initialized, your python process automatically connect to the ray cluster,
+    you can call `ray.util.spark.shutdown_ray_cluster` to shut down the ray cluster.
+    Note: If the active ray cluster haven't shut down, you cannot create a new ray cluster.
+
+    Args
+        num_worker_nodes: The number of spark worker nodes that the spark job will be submitted to.
+            This argument represents how many concurrent spark tasks will be available in the
+            creation of the ray cluster. The ray cluster's total available resources
+            (memory, CPU and/or GPU)
+            is equal to the quantity of resources allocated within these spark tasks.
+            Specifying the `num_worker_nodes` as `-1` represents a ray cluster configuration that
+            will use all available spark tasks slots (and resources allocated to the spark
+            application) on the spark cluster.
+            To create a spark cluster that is intended to be used exclusively as a shared ray
+            cluster, it is recommended to set this argument to `-1`.
+        total_cpus: The total cpu core count for the ray cluster to utilize.
+        total_gpus: The total gpu count for the ray cluster to utilize.
+        total_heap_memory_bytes: The total amount of heap memory (in bytes) for the ray cluster
+            to utilize.
+        total_object_store_memory_bytes: The total amount of object store memory (in bytes) for
+            the ray cluster to utilize.
+        heap_to_object_store_memory_ratio: The ratio of per-ray worker node available memory to the
```

Comment:
> based on /dev/shm free space

On many aws instances, the /dev/shm is more than half of physical memory size, is the ratio too high ? The option is for allowing user to controlling the ratio.

Reply:
How about renaming this to something like `object_store_fraction` then? It can be documented as an advanced option, by default 0.8 (and only reduce this if you are sure about it). Right now, it's quite hard to understand.

Reply:
Double confirm: object_store_fraction by default 0.8, means by default 80% memory will be used in object store, and 20% will be used as ray tasks heap memory, is that you expect ?

Reply:
Not exactly. I think in the current code you are calculating object store memory already as `= 0.8 * dev_shm_avail_size / num_tasks_per_node`. I was suggesting just replacing that hard-coded 0.8 with this fraction, so `= object_store_fraction * dev_shm_avail_size / num_tasks_per_node`.

Another alternative I can think of is to specify instead just `object_store_memory_per_worker` directly, raising an exception if it exceeds `dev_shm_avail_size / num_tasks_per_node`. Not sure if that would be more intuitive for users.

Reply:
@ericl 
I like `object_store_memory_per_worker` , but shall we rename it as `object_store_memory_per_node` ? 
The "worker" in ray seemingly usually means "a worker running a ray task" but ray node contains multiple workers. To avoid confusion, `object_store_memory_per_node` looks better